### ISSUE TITLE
Error: Code="InvalidResourceType" Message="The resource type 'operationResults/asyncOperation' could not be found in the namespace 'Microsoft.Management' for api version '2018-03-01-preview'. The supported api-versions are '2021-04-01,2020-10-01,2020-05-01,2020-02-01,2019-11-01,2018-03-01-beta'."

### ISSUE THREAD

**Issue by riteshmehta66** [Azure/terraform-azurerm-caf-enterprise-scale#239]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Versions

**terraform**: v1.0.4

**azure provider**:  2.89.0 ( terraform.tf files has version = ">= 2.77.0" )

**module**:  1.0.0

### Description

**Describe the bug**

module.enterprise_scale.azurerm_management_group.level_1["/providers/Microsoft.Management/managementGroups/mhb-dev-rt"]: Still creating... [3m21s elapsed]
module.enterprise_scale.azurerm_management_group.level_1["/providers/Microsoft.Management/managementGroups/mhb-dev-rt"]: Still creating... [3m31s elapsed]
module.enterprise_scale.azurerm_management_group.level_1["/providers/Microsoft.Management/managementGroups/mhb-dev-rt"]: Still creating... [3m41s elapsed]
╷
│ Error: failed when waiting for creation of Management Group "mhb-dev-rt": Future#WaitForCompletion: the number of retries has been exceeded: StatusCode=404 -- Original Error: Code="InvalidResourceType" Message="The resource type 'operationResults/asyncOperation' could not be found in the namespace 'Microsoft.Management' for api version '2018-03-01-preview'. The supported api-versions are '2021-04-01,2020-10-01,2020-05-01,2020-02-01,2019-11-01,2018-03-01-beta'."
│
│   with module.enterprise_scale.azurerm_management_group.level_1["/providers/Microsoft.Management/managementGroups/mhb-dev-rt"],
│   on .terraform\modules\enterprise_scale\resources.management_groups.tf line 9, in resource "azurerm_management_group" "level_1":
│    9: resource "azurerm_management_group" "level_1" {
│
╵

**Steps to Reproduce**

1. kindly let me know your github Id. I can add you to my repo to have the config files.
2. step 2
3. you get it...

**Screenshots**

**Additional context**

**Comment by krowlandson** [comment]
@riteshmehta66... are you still able to replicate this issue?

This error looks like a provider or SDK issue, as we do not (and cannot) set the API versions from within the module. I am also unable to replicate this issue, so although the Azure provider `v2.89.0` does appear to use api version '2018-03-01-preview' as mentioned in the error, this works fine when I try to reproduce the problem.

**Comment by riteshmehta66** [comment]
yes, still an issue. Only timeout changes. in some cases get error after 10 seconds and in some case after couple of minutes.  I tried setting version= "= 2.88.1" and checked C:\Data\IaC\Terraform\caf-tf-enterprise-scale\root\configuration\.terraform\providers\registry.terraform.io\hashicorp\azurerm\2.88.1\windows_amd64\terraform-provider-azurerm_v2.88.1_x5.exe is downloaded. I am still getting error

Plan: 81 to add, 0 to change, 0 to destroy.
module.enterprise_scale.azurerm_management_group.level_1["/providers/Microsoft.Management/managementGroups/mhb-dev-rt"]: Creating...
module.enterprise_scale.azurerm_management_group.level_1["/providers/Microsoft.Management/managementGroups/mhb-dev-rt"]: Still creating... [10s elapsed]
╷
│ Error: failed waiting for read on Managementgroup "mhb-dev-rt"
│
│   with module.enterprise_scale.azurerm_management_group.level_1["/providers/Microsoft.Management/managementGroups/mhb-dev-rt"],
│   on .terraform\modules\enterprise_scale\resources.management_groups.tf line 9, in resource "azurerm_management_group" "level_1":
│    9: resource "azurerm_management_group" "level_1" {
│

=========================================================================

I also tried changing to version = "= 2.80" but getting similar error:

    }

Plan: 81 to add, 0 to change, 0 to destroy.
module.enterprise_scale.azurerm_management_group.level_1["/providers/Microsoft.Management/managementGroups/mhb-dev-rt"]: Creating...
module.enterprise_scale.azurerm_management_group.level_1["/providers/Microsoft.Management/managementGroups/mhb-dev-rt"]: Still creating... [10s elapsed]
╷
│ Error: failed waiting for read on Managementgroup "mhb-dev-rt"
│
│   with module.enterprise_scale.azurerm_management_group.level_1["/providers/Microsoft.Management/managementGroups/mhb-dev-rt"],
│   on .terraform\modules\enterprise_scale\resources.management_groups.tf line 9, in resource "azurerm_management_group" "level_1":
│    9: resource "azurerm_management_group" "level_1" {
│
╵

**Comment by matt-FFFFFF** [comment]
Hi @riteshmehta66,

I have also found this issue outside this module. It looks like it's a problem upstream, with either the provider (which uses API version `2018-03-01-preview` for management groups), or the ARM resource provider itself.

I'm going to close this as it's not specific to this module if that's ok?

**Comment by khabbabs** [comment]
Getting the exact same issue.

**Comment by matt-FFFFFF** [comment]
I have found that this is a transient error. Although not ideal, retrying the operation has yielded success (for me anyway).

**Comment by khabbabs** [comment]
unfortunately, retries haven't been successful for me.

**Comment by Brudolph007** [comment]
i have also experienced these issues. I have tried to use previous providers >=2.80. retires, re-log into the tenant also does not help. 
Thank you.

**Comment by khabbabs** [comment]
similar issue from the azurerm github issues. https://github.com/hashicorp/terraform-provider-azurerm/issues/14558

Could this be related to the log4j remediations ?

**Comment by matt-FFFFFF** [comment]
This seems to be more widespread now. We are chasing internally.

**Comment by krowlandson** [comment]
Pinned issue as this appears to be a broader issue across Azure. Not just Terraform clients being impacted.

**Comment by kevinevans1** [comment]
I'm receiving this issue when trying to deploy as well current Terraform azurerm provider version 2.77.0

Error: failed when waiting for creation of Management Group "dc": Future#WaitForCompletion: the number of retries has been exceeded: StatusCode=404 -- Original Error: Code="InvalidResourceType" Message="The resource type 'operationResults/asyncOperation' could not be found in the namespace 'Microsoft.Management' for api version '2018-03-01-preview'. The supported api-versions are '2021-04-01,2020-10-01,2020-05-01,2020-02-01,2019-11-01,2018-03-01-beta'."
│ 
│   with module.enterprise_scale.azurerm_management_group.level_1["/providers/Microsoft.Management/managementGroups/dc"],
│   on .terraform/modules/enterprise_scale/resources.management_groups.tf line 9, in resource "azurerm_management_group" "level_1":
│    9: resource "azurerm_management_group" "level_1" {
│

**Comment by evanbrothers** [comment]
Receiving this issue as well with the same error detailed by @kevinevans1 . Any updates on this?  It's literally blocking our entire management group creation solution.

**Comment by matt-FFFFFF** [comment]
> Receiving this issue as well with the same error detailed by @kevinevans1 . Any updates on this?  It's literally blocking our entire management group creation solution.

Hi,

We don't have anything more to share at the time. We realise this is frustrating - we are also blocked on our end to end testing.

We have escalated this internally and been in touch with relevant teams, but do not have an ETA at the moment.

**Comment by a-illarionov** [comment]
I am also getting this exact issue, it will work sometimes, but 95% of the time it errors out. 

Are there any work arounds that have been figured out yet?

**Comment by jtracey93** [comment]
Hey @a-illarionov,

As mentioned above this is an upstream issue and is impacting more than just this module, and Terraform for that matter. 

We are working with teams internally to work towards a resolution but do not have any timescales as of yet.

**Ask for everyone impacted by this issue**: please create an Azure support ticket as well as posting here, as this will likely be very useful for us and yourselves for getting this resolved ASAP 👍

Thanks 

Jack

**Comment by krowlandson** [comment]
The following PR is expected to resolve this issue:
https://github.com/hashicorp/terraform-provider-azurerm/pull/14635

Please note that this will require **ALL** customers (regardless of use of this module) who deploy Management Groups using Terraform to update to the latest provider version.

We are continuing to investigate the root cause and see whether the original API version can be fixed/re-instated due to the broad impact this will have on customers using Terraform.

**Comment by matt-FFFFFF** [comment]
### Update

We have confirmed that the management groups engineering team is looking at this and has identified the issue. A fix for the `2018-03-01-preview` api is rolling out but we do not have timescales at the moment.

I will report back with more information when we have it.

**Comment by matt-FFFFFF** [comment]
### Workaround

The latest Terraform provider (v2.90.0) contains an updated management groups API version that does not suffer from this issue.

**Comment by matt-FFFFFF** [comment]
@riteshmehta66 Please can you confirm this is fixed?

**Comment by kevinevans1** [comment]
running test deployment now....

**Comment by kevinevans1** [comment]
Just completed a successful deployment after upgrading to latest azurerm provider in Terraform. Managment groups deployed successfully and Contoso reference architecture deployed across multiple subscriptions.

Once you added the latest azurerm provider to you configuration "2.90." run "terraform init -upgrade"    

![image](https://user-images.githubusercontent.com/68919619/146535976-bb9606d7-30ad-4272-b9a3-992b486876b1.png)

@matt-FFFFFF @krowlandson @jtracey93 @tombuildsstuff @a-illarionov @evanbrothers @khabbabs

**Comment by tohov** [comment]
Hmm...we're running TF version 1.0.5 and azurerm provider "2.90.0" but still cannot get this to work.

**Comment by kevinevans1** [comment]
> Hmm...we're running TF version 1.0.5 and azurerm provider "2.90.0" but still cannot get this to work.

My tests where completed with the following Terraform versions:
Terraform v1.1.1
on linux_amd64
+ provider registry.terraform.io/hashicorp/azurerm v2.90.0
+ provider registry.terraform.io/hashicorp/time v0.7.2

**Comment by jtracey93** [comment]
Hey @tohov, we are also seeing this and are speaking with engineering teams to advise them.

@kevinevans1 would you mind testing again for us as we think it may of changed between now and when you tested last. This would help us confirm 👍

Apologies for this, please do raise a support ticket on your azure subscription @tohov to help us highlight this internally👍

**Comment by kevinevans1** [comment]
> Hey @tohov, we are also seeing this and are speaking with engineering teams to advise them.
> 
> @kevinevans1 would you mind testing again for us as we think it may of changed between now and when you tested last. This would help us confirm 👍
> 
> Apologies for this, please do raise a support ticket on your azure subscription @tohov to help us highlight this internally👍

Running test deployment now 👍

**Comment by tohov** [comment]
We don't have the permissions to create support ticket so we'll have to contact the customer who has permissions next week and create it then. Thanks for the help.

**Comment by kevinevans1** [comment]
Deployment successful :) @jtracey93 

![image](https://user-images.githubusercontent.com/68919619/146582083-5d1c4924-cf09-405e-9e53-f3e00720569d.png)

**Comment by jtracey93** [comment]
Thanks @kevinevans1. My testing also shows provider version2.90.0 working
![image](https://user-images.githubusercontent.com/41163455/146591222-8bdee459-2574-4358-9926-408ef9c84277.png)

However, version 2.89.0 is still failing but the MGs do now create:
![image](https://user-images.githubusercontent.com/41163455/146591321-5d8a16a5-5785-464b-88f1-b7b6185cac1e.png)

I have been speaking with engineering this evening and they are looking into this further 👍

Will keep you all posted, but upgrading to provider version 2.90.0 still seems to be the best path forward at the moment 👍

**Comment by riteshmehta66** [comment]
> @riteshmehta66 Please can you confirm this is fixed?

Hello @matt-FFFFFF - Yes, I could deploy Management Groups with latest azurerm provider version 2.90. Terraform version 1.04, module version 1.0.0.

Though this may not be related with earlier error... I am getting timeout error as shown below. If I run terraform apply again , completes successfully. Any idea on this ? I can create new issue to keep it separate.

==========
 
module.enterprise_scale.azurerm_management_group_policy_assignment.enterprise_scale["/providers/Microsoft.Management/managementGroups/17dec738pm-lz-dev-byn-dev/providers/Microsoft.Authorization/policyAssignments/xxx-dev-search-pol"]: Creation complete after 1m38s [id=/providers/Microsoft.Management/managementGroups/17dec738pm-lz-dev-byn-dev/providers/Microsoft.Authorization/policyAssignments/xxx-dev-search-pol]
module.enterprise_scale.azurerm_management_group_policy_assignment.enterprise_scale["/providers/Microsoft.Management/managementGroups/17dec738pm-lz-dev-byn-dev/providers/Microsoft.Authorization/policyAssignments/xxx-dev-backup-vm-pol"]: Creation complete after 1m37s [id=/providers/Microsoft.Management/managementGroups/17dec738pm-lz-dev-byn-dev/providers/Microsoft.Authorization/policyAssignments/xxx-dev-backup-vm-pol]
╷
│ Error: creating Policy Assignment ID: (Assignment Name "xxx-dev-app-pol" / Scope "/providers/Microsoft.Management/managementGroups/17dec738pm-lz-dev-byn-dev"): policy.AssignmentsClient#Create: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="InvalidCreatePolicyAssignmentRequest" Message="The policy definition specified in policy assignment 'xxx-dev-app-pol' is out of scope. Policy definitions should be specified only at or above the policy assignment scope. If the management groups hierarchy changed recently or if assigning a management group policy to new subscription, please allow up to 30 minutes for the hierarchy changes to apply and try again."
│
│   with module.enterprise_scale.azurerm_management_group_policy_assignment.enterprise_scale["/providers/Microsoft.Management/managementGroups/17dec738pm-lz-dev-byn-dev/providers/Microsoft.Authorization/policyAssignments/xxx-dev-app-pol"],
│   on .terraform\modules\enterprise_scale\resources.policy_assignments.tf line 1, in resource "azurerm_management_group_policy_assignment" "enterprise_scale":
│    1: resource "azurerm_management_group_policy_assignment" "enterprise_scale" {
│
│ creating Policy Assignment ID: (Assignment Name "xxx-dev-app-pol" / Scope "/providers/Microsoft.Management/managementGroups/17dec738pm-lz-dev-byn-dev"): policy.AssignmentsClient#Create: Failure
│ responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="InvalidCreatePolicyAssignmentRequest" Message="The policy definition specified in    
│ policy assignment 'xxx-dev-app-pol' is out of scope. Policy definitions should be specified only at or above the policy assignment scope. If the management groups hierarchy changed recently or if       
│ assigning a management group policy to new subscription, please allow up to 30 minutes for the hierarchy changes to apply and try again."
╵
╷

**Comment by krowlandson** [comment]
@riteshmehta66, unfortunately this is an issue in the Management Groups resource provider (on the Azure side) related to replication of changes, so nothing we can do about this. This is why we have a number of configurable `time_sleep` resources in the module to help with the above, but in rare cases the default values are not quite enough so a re-run is required as you observed.

**Comment by MitchPrince-HiSol** [comment]
The above issue creating management groups was resolved for me by upgrading to the azurerm provider 2.90.0 that came out a few days ago.  There's a fix that updated managementgroups to API Version 2020-05-01 ([#14635](https://github.com/hashicorp/terraform-provider-azurerm/issues/14635))

**Comment by krowlandson** [comment]
Telemetry suggests that the underlying problem is now resolved so will close this issue.
